### PR TITLE
Client should properly validate match commit checksum upon match_proof message.

### PR DESCRIPTION
**Pull request by LasTshaMAN**
I didn't try to deploy DEX locally and test this change manually (yet), do we have some basic automatic smoke tests that would uncover issues (if any) with changes similar to these ?

Closes https://github.com/decred/dcrdex/issues/1020.

### CODE REVIEW

File: client/core/trade.go
Comment (LasTshaMAN) [inline client/core/trade.go:186]: this is dead code, so, removing it

File: dex/msgjson/types.go
Comment (LasTshaMAN) [inline dex/msgjson/types.go:832]: Is it okay to change this type ? Changing this type seems only natural to me, to avoid unnecessary type-conversions when finding relevant tracked orders in `validateMatchCsum` func.

Comment (chappjc) [reply dex/msgjson/types.go:832]: Depends on how it marshals/unmarshals to JSON.  I don't believe it implements either json.Marshaller or Unmarshaller, so I suspect it's going to be an array of integers as a result instead of hex.

We lack a TestMatchProof in msg_test, but you could make one to see if it marshals and unmarshals as expected.

```go
func TestMatchProof(t *testing.T) {
	pimgB := randomBytes(32)
	var pimg order.Preimage
	copy(pimg[:], pimgB)
	match := &MatchProofNote{
		Preimages: []order.Preimage{pimg},
	}

	matchB, err := json.Marshal(match)
	if err != nil {
		t.Fatalf("marshal error: %v", err)
	}

	t.Logf("%v", string(matchB))

    // check things here

	var matchBack Match
	err = json.Unmarshal(matchB, &matchBack)
	if err != nil {
		t.Fatalf("unmarshal error: %v", err)
	}

    // check more things here
}
```

> === RUN   TestMatchProof
>     /home/jon/github/decred/dcrdex/dex/msgjson/msg_test.go:83: {"marketid":"","epoch":0,"preimages":[[82,253,252,7,33,130,101,79,22,63,95,15,154,98,29,114,149,102,199,77,16,3,124,77,123,187,4,7,209,226,198,73]],"misses":null,"csum":"","seed":""}
> 

With `[]Bytes`:

> === RUN   TestMatchProof
>     /home/jon/github/decred/dcrdex/dex/msgjson/msg_test.go:80: {"marketid":"","epoch":0,"preimages":["52fdfc072182654f163f5f0f9a621d729566c74d10037c4d7bbb0407d1e2c649"],"misses":null,"csum":"","seed":""}

File: client/core/trade.go
Comment (chappjc) [reply client/core/trade.go:186]: My linters have caught that too.  Thanks.

File: client/core/core_test.go
Comment (chappjc) [inline client/core/core_test.go:2531]: Deadlock if it does hit t.Fatal.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:5034]: Changes to product code just to facilitate test, particularly blocking calls, is to be avoided like the plague.

If the test gets the `SubjectPreimageSent` note emitted below, we can be sure that this code has already executed and there's no risk of a data race.  Was a race detected?

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:907]: I know this gripe seems silly, but it's really not an burden to just setup your editor with a key combo to automatically wrap selected text.  It takes virtually no thought, it makes docs look consistent, and we all do it unless the content of the comment demands an exception.  This looks super janky right now.

Comment (JoeGruffins) [reply client/core/core.go:907]: I just use `:set colorcolumn=80` with vim and do it manually though that's a little "neckbeard"ish.

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go:5034]: > Changes to product code just to facilitate tests

I agree, but this has nothing to do with tests, 

`tracker.csum` gets written to in one request (go-routine 1), while we a reading them in another request (go-routine 2),

maybe due to the time difference between these 2 requests it is safe to assume that `tracker.csum` will be synchronized, but its not 100% the case, so I'm just playing it safe here,
if this causes any performance issues, then we can probably omit locking here,

> Was a race detected?

nope, there is syncronization on channel that prevents race, given its current implementation

File: client/core/core.go
Comment (buck54321) [inline client/core/core.go:908]: This name change adds nothing and is completely unrelated to the task at hand. In general, do not change anything except what's necessary. We will sometimes accept an un-related or stylistic change as part of PR, but not until you've established that you can keep your work properly scoped.

File: client/core/core.go
Comment (buck54321) [inline client/core/core.go:4605]: This is going to error virtually every time. The server is sending every preimage of every order from the epoch, not just our own.

File: client/core/core.go
Comment (buck54321) [inline client/core/core.go:4616]: Don't check every trade, only those in `OrderStatusEpoch` for which we've received a preimage request.

File: client/core/core.go
Comment (buck54321) [inline client/core/core.go:4582]: The use of the two helper methods is overkill. Just iterate here.

```
for _, trade := range dc.trackedTrades() {
    if trade.metaData.Status == order.OrderStatusEpoch && len(trade.csum) > 0 && !bytes.Equal(note.CSum, trade.csum)
        // Probably self-revoke the order so we don't acknowledge match requests. Also log the error.
    }
}
```

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go:907]: Changed formatting for this section.

I'm formatting everything manually at the moment the way that seems to be the most readable to me, but looks like it's not working as expected :)

Long-term, I should probably setup something to help me. Other comments look OK ?

File: client/core/core_test.go
Comment (LasTshaMAN) [reply client/core/core_test.go:2531]: true, although I believe `TestHandlePreimageRequest` to be a pretty simple isolated test case with nothing vital here being locked forever,

when we hit `t.Fatal` here, all the related stuff around `tracker.mtx` becomes useless,

but maybe it is better to just copy value, re-wrote this part

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go:4604]: hmm, good point, haven't realized that `trackedTrade` represents an order issued by "this client only" and not just any order with additional meta,

shall I modify a description for `trackedTrade` ?

```
// trackedTrade is an order, its matches, and its cancel order, if applicable.
// The trackedTrade has methods for handling requests from the DEX to progress
// match negotiation.
```
to this
```
// trackedTrade is an order (issued by this client), its matches, and its cancel
// order, if applicable. The trackedTrade has methods for handling requests
// from the DEX to progress match negotiation.
```

as well as add a comment for `dexConnection.trades` to clarify that ?

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go:907]: Right, should I remove this change ?

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go:4581]: What's the purpose for this check `len(trade.csum) > 0` ? 

My understanding is - if `len(trade.csum) == 0` this validation must fail as well.

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go]: fixed.

File: client/core/core.go
Comment (chappjc) [reply client/core/core.go:5010]: We shouldn't roll the dice with data races based on general ideas about expected timing, no.

I'm commenting about the fact that in the path in `handleMatchProofMsg` you aren't locking the mutex for the `csum` read, only in the test and here in `processPreimageRequest`.  Hence it looks like you've set up the locking for the test.

But here's the sequence:
- when an epoch closes, server computes the csum from all order commitments in that epoch, and it sends preimage requests to clients for all those orders
- client gets a preimage request, starts `processPreimageRequest`, which records the csum that came with the request and **then** responds to the server with the preimage. So server only has the preimage if the csum is recorded.
- when the server gets all the epoch preimages (or times out on any of them), it performs matching with those it has preimages for
- server sends a match_proof note, which contains the preimages **that it received (obviously)** and the csum it computed from all the epoch order commitments
- `handleMatchProofMsg` looks (or should) for orders with the corresponding preimages.  If the a known preimage is found, that's because `processPreimageRequest` has completed already.

So if the preimage for the order is in the match_proof note, that means the server has received it from `processPreimageRequest` and the tracker.csum has already been saved.  No race.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:4597]: To start I only want to warn.  We've done nothing so far, so we should at least use this in production with just a warning until we're sure we're not going to be bailing out on orders incorrectly, which will cause swaps to fail and users to be penalized (account suspended)

File: client/core/core.go
Comment (chappjc) [reply client/core/core.go:908]: Yeah, pls leave `setEpoch` and it's docs as-is. Thanks.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:4585]: Since epochs are continuous and there is time between epoch close and this match_proof msg (up to 20 sec for the preimage request timeout + network latency, although the spec incorrectly says 5 sec), there can be trades in `OrderStatusEpoch` from more than one epoch (different csums).

I'm afraid the `trackedTrade` doesn't know what epoch index it belongs to (sadly, server's `OrderResponse` doesn't include that), so finding the order with the corresponding preimage may be needed after all. However, that facilitates the lock-free solution to the `csum` data race issue.  I may be overlooking something...  @buck54321 I'm thinking cycling over preimages is gonna be needed after all, although I agree it doesn't quite warrant 2 helper functions.

As an aside, I'm really not sure if `dc.epoch` (and `setEpoch`) serves a purpose anymore, but let's not go down that rabbit how right now.  There are relics here and there we should note, but not change too readily.

File: client/core/core_test.go
Comment (chappjc) [reply client/core/core_test.go]: I get what you're saying, but it could hang at 

```
defer rig.shutdown()
````

`Core` might be unable to stop depending on what it's doing at the time.  I've had tests hang on t.Fatals before and it's very confusing so best to play it safe.  Thanks for changing.

File: client/core/core.go
Comment (chappjc) [reply client/core/core.go]: I commented in https://github.com/decred/dcrdex/pull/1077#discussion_r627812050, but I actually think we have to look at all of our trades' preimages because OrderStatusEpoch orders can be from multiple epochs.  If trackedTrade knew which epoch index it was in, that would work too, alas it does not know.

File: client/core/core.go
Comment (chappjc) [reply client/core/core.go]: That would imply dexc has restarted and csum was not restored (a rare event as restarting with epoch orders is a super dumb idea), or the trade is from a different epoch and a preimage request has not been received for that trade.  Related to my other comments regarding orders in OrderStatusEpoch potentially spanning multiple epochs.

File: client/core/core.go
Comment (JoeGruffins) [inline client/core/core.go:120]: Srry for nitpick, but period at the end. I treat `trades` here as a proper noun although not capitalized, so grammatically correct to call it a sentence, with that one exception. I think.

File: client/core/core.go
Comment (JoeGruffins) [reply client/core/core.go:908]: Either way is fine with me... but another pr, even for just this one change is probably better.

File: client/core/core.go
Comment (JoeGruffins) [inline client/core/core.go:4568]: I do like the period here!

**Review by JoeGruffins — COMMENTED** [review]
Running the simnet tests:
```
2021-05-07 09:54:15.016 [INF] MKT: Book router terminating for market "dcr_btc"
panic: runtime error: index out of range [0] with length 0

goroutine 135 [running]:
decred.org/dcrdex/server/market.(*BookRouter).runBook(0xc0005af050, 0xd60df8, 0xc000576d00, 0xc000347500)
        /home/joe/git/dcrdex/server/market/bookrouter.go:455 +0x7d5
decred.org/dcrdex/server/market.(*BookRouter).Run.func1(0xc0005af050, 0xd60df8, 0xc000576d00, 0xc0000256d0, 0xc000347500)
        /home/joe/git/dcrdex/server/market/bookrouter.go:312 +0x49
created by decred.org/dcrdex/server/market.(*BookRouter).Run
        /home/joe/git/dcrdex/server/market/bookrouter.go:311 +0x10e
```

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go:5010]: Okay, I see ... Turns out `preimage` and `match_proof` requests are both websocket requests that get through the same channel (thus it makes any inmemory state between them synchronized).

So, there is no need to synchronize explicitely.

Leaving here my flawed (due to lack of context) reasoning just for completeness.

<details>
<summary>my flawed reasoning</summary>
<p>

I understand the sequence you've described, but not sure what you mean by "saved" (`trackedTrade`s reside in process memory between `preimage` and `match_proof` requests) in "and the tracker.csum has already been saved. No race.".

> I'm commenting about the fact that in the path in handleMatchProofMsg you aren't locking the mutex for the csum read, only in the test and here in processPreimageRequest.

That's a very good point, and I didn't look at it from this perspective, let me clarify.

The way it is implemented right now, it looks indeed as if locking on `tracker.mtx` here is done purely for testing purposes. But it not the intent. 
The intent is - for any other go-rotine G that is interested in reading up-to-date `tracker.csum` value (or modifying it), the developer implementing G will only have to worry about aquiring/releasing this lock in G properly, and not follow through the code checking whether `tracker.csum` was written to with this lock being used or not.
This is the contract that, I assumed, everyone should follow when working with `trackedTrade` fields, protected by this lock (based on its description):
```go
	// mtx protects all read-write fields of the trackedTrade and the
	// matchTrackers in the matches map.
	mtx      sync.RWMutex
```
To summarize, we aquire this lock here in order to not shoot ourselves in the foot later on when implementing features that depend on `tracker.csum` value. 
Whether this opens a whole lot of different issues to worry about, such as deadlocks or bad performance, is another matter also worthy to think through.

Now, back to `handleMatchProofMsg`, why didn't I lock on `tracker.mtx` here ? This is my bad indeed, looks like we need to lock it there as well ...
</p>
</details>

File: dex/msgjson/types.go
Comment (LasTshaMAN) [reply dex/msgjson/types.go]: Rolled back type change.

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go:908]: Will move this to a separate PR.

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go:4597]: Okay, warning it is for now.

**Comment by LasTshaMAN** [comment]
> Running the simnet tests:
> 
> ```
> 2021-05-07 09:54:15.016 [INF] MKT: Book router terminating for market "dcr_btc"
> panic: runtime error: index out of range [0] with length 0
> 
> goroutine 135 [running]:
> decred.org/dcrdex/server/market.(*BookRouter).runBook(0xc0005af050, 0xd60df8, 0xc000576d00, 0xc000347500)
>         /home/joe/git/dcrdex/server/market/bookrouter.go:455 +0x7d5
> decred.org/dcrdex/server/market.(*BookRouter).Run.func1(0xc0005af050, 0xd60df8, 0xc000576d00, 0xc0000256d0, 0xc000347500)
>         /home/joe/git/dcrdex/server/market/bookrouter.go:312 +0x49
> created by decred.org/dcrdex/server/market.(*BookRouter).Run
>         /home/joe/git/dcrdex/server/market/bookrouter.go:311 +0x10e
> ```

This is due to a bug (with slice creation) I've introduced when changing `Bytes` -> `order.Preimage` type, as disscussed here - https://github.com/decred/dcrdex/pull/1077#discussion_r626638772

image

rolled back this type change, hopefully its fixed now.

**Comment by LasTshaMAN** [comment]
I've re-worked this PR trying to address all the comments, and closed off all discussions that became irrelevant due to all the latest changes.

Waiting for @buck54321's response to @chappjc's concerns about `server epoch` vs `order.OrderStatusEpoch` mismatch before modifying anything further.

File: client/core/core.go
Comment (buck54321) [reply client/core/core.go]: No need. It doesn't improve clarity or readability, and even if it did, it's not something to spend time on right now.

File: client/core/core.go
Comment (chappjc) [reply client/core/core.go]: I'm honestly pleased that you're attuned to the need to synchronize with concurrent writes, but no need to go into it much - we're all on the same page about the need for that.

Yes, there is implicit synchronization that prevents concurrent writes, but it's not because "both websocket requests that get through the same channel" if I understand your meaning.  It's because a match_proof request can not include the preimage if `processPreimageRequest` has not already "saved" the csum (`tracker.csum = commitChecksum`). The other goroutine that reads `tracker.csum` will only attempt that read if the server knew about that preimage to include it with the match_proof note, which means it has already been stored in `tracker.csum`.  This all assumes the first iteration of these changes where preimages were checked to identify relevant trades.


Now a comment about the following:

> developer implementing G will only have to worry about aquiring/releasing this lock in G properly, and not follow through the code checking whether tracker.csum was written to with this lock being used or not.

You see this as being safe, but in reality not understanding access patterns and instead falling back to just locking mutexes liberally is seriously problematic.  When possible, it is generally wiser to design access patterns that do not involve the possibility for concurrent writes.  You'll have to believe me when I say it can become very messy with functions that maybe lock and maybe don't, called from other methods that do lock... so is it gonna deadlock or is it gonna be a race or is it gonna be correct?  This is particularly easy to get into with `trackedTrade.RWMutex` given how it's used so extensively in Core.

In short, the `csum` field should not be under the `mtx` in the struct definition because: it will only ever be accessed AFTER it is written, if the access patterns design for that.

Comment (chappjc) [reply client/core/core.go]: Although I've neglected the possibility of a malfunctioning server sending multiple preimage requests for the same order/commitment... :sigh:  I'm considerably less confident that we'd not end up with two concurrent writes (two `processPreimageRequest`).  We'd also want logic to reject such requests if `csum` is already stored.  Let's first see how this plays out after buck weighs in on my assertion that there can be epoch status orders for multiple epochs.

We might end up needing a sync mechanism for the `csum` field after all.  If that is the case we can consider it's own mutex or something more low-level like an atomic uint32 CAS that just says if it has been stored already so that we don't have to worry about an locking at the read sites.

File: client/core/core.go
Comment (buck54321) [reply client/core/core.go]: > If trackedTrade knew which epoch index it was in

As long as we have the `order.Prefix.ServerTime` and the `msgjson.Market.EpochLen`, we have the epoch index, and we can match against `msgjson.MatchProofNote.Epoch`.

Comment (buck54321) [reply client/core/core.go]: Yep, which can simplify the filter too.

```
epochIdx := encode.UnixMilliU(trade.Prefix().ServerTime) / trade.epochLen
if note.Epoch == epochIdx && note.MarketID == trade.mktID {
    // do stuff
}
```

Note the addition of the market ID comparison, which I had failed to notice before.

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go]: Thanks to the detailed explanations above, I think now we are on the same page regarding:
- related concurrency considerations
- we must assume the server is maliscious

Below are some conclusions I've arrived at, please let me know wheter any of these are incorrect.

1. Checking `len(trade.cSum) > 0` in `handleMatchProofMsg` is probably not sufficient, because maliscious server might just provide empty value for `CommitChecksum` in `handlePreimageRequest` resulting in client omitting `cSum` validation altogether.

2. I believe we shouldn't check for duplicate `handlePreimageRequest` requests from the server, instead we should apply every request (overwriting the previous `cSum` value) and just verifying the value matches should be enough to spot any maliscious behavior by the server. This looks like a simple enough solution, given (3).

3. 
> but that might warrant resetting it to 0 if the websocket Send just fails due to connectivity or whatever so a subsequent request can be fielded.

Not sure how to better handle that ... because it might be difficult to differentiate between "server playing games" & "connectivity issues".

**Review by JoeGruffins — APPROVED** [review]
Looks good to me.

File: client/core/core.go
Comment (JoeGruffins) [inline client/core/core.go:4583]: Is it possible for `processPreimageRequest` to have not happened yet?

I don't see a check for `nil` in `processPreimageRequest`, unless I am missing a check somewhere else. Maybe server can send nil in the `PreimageRequest` request?

File: client/core/core.go
Comment (JoeGruffins) [inline client/core/core.go:4938]: I just tried irl and it seems passing a nil `req.CommitChecksum` is fine atm. Maybe we want to error (or just warn for now) before this if the check sum is nil or wrong length? I'm guess up around the check for `req.OrderID` length.

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go:4583]: Yes, I've mentioned that in my comment above (section 3), we should probably check against that,

https://github.com/decred/dcrdex/pull/1077#discussion_r628744407

Comment (JoeGruffins) [reply client/core/core.go:4583]: Ah, somehow I missed that long conversation. Well, I don't know the best answer, but thank you for your time on this. imo coming along nicely!

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go]: @chappjc @buck54321 any comments on my summary above ?

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:5007]: We've forgotten about cancel orders, which are not `trackedTrades` themselves, but the `trackedTrade.cancel` field that links it with the targeted trade.  Please see the following suggested diff: https://github.com/decred/dcrdex/commit/cbf49a8512913b4683e611e2d1c272a068b9e64f

I'd also prefer the `trackedTrade` formatted like in that diff too if you don't mind applying that too.

Comment (LasTshaMAN) [reply client/core/core.go]: Okay, I've applied your diff, thanks! Looks like it addresses all the concerns I've outlined above - https://github.com/decred/dcrdex/pull/1077#discussion_r628744407

File: client/core/core.go
Comment (chappjc) [reply client/core/core.go]: > I believe we shouldn't check for duplicate handlePreimageRequest requests from the server, instead we should process every request (overwriting the previous csum value) and just verifying the value matches should be enough to spot any maliscious behavior by the server.

On the contrary, allowing duplicate requests to essentially update the commitment checksum is facilitating the server cheating.  Server would just have to send a final preimage request just before the match_proof ntfn with a csum that matches, thus defeating the check in handleMatchProof.

If the trade/cancel's `csum` is already set and not equal to what the repeated request is providing, that's definitely server cheating and should be handled as such.

Comment (chappjc) [reply client/core/core.go]: Also, if you could place a TODO comment in the test for handleMatchProof, we can get to that when we are ready to change these warnings into errors that prevent client from negotiating matches for such orders.

File: client/core/core.go
Comment (chappjc) [reply client/core/core.go]: I should just note as a follow-up to the other discussion in this PR w.r.t. mutex/atomic that I felt using the `trackedTrade.mtx` was alright in the end because we had to safely access the `cancel` field as well.

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go]: Added a `TODO` comment about handling duplicates (just a comment, no code to handle that for now).

**Review by buck54321 — CHANGES_REQUESTED** [review]
This is really looking good now. A little test coverage to round out, but otherwise seems ready to me. Thanks for sticking it out.

File: client/core/core_test.go
Comment (LasTshaMAN) [reply client/core/core_test.go:2537]: will do)

Comment (LasTshaMAN) [reply client/core/core_test.go:2537]: Moved `csum` related checks into separate tests, since trying to tests everything in a single huge test is pretty tricky and unreadable.

**Review by chappjc — APPROVED** [review]
OK.

File: client/core/core.go
Comment (chappjc) [reply client/core/core.go]: I've approved the PR, but if @buck54321 has anything else that requires a commit, I'd really like to have `processPreimageRequest` return an error if the `tracker.{cancel.}csum` is non-nil.  Otherwise these changes offer no real protection yet, just a start in that direction.

**Review by buck54321 — APPROVED** [review]

File: client/core/core.go
Comment (LasTshaMAN) [reply client/core/core.go]: @chappjc I can create a follow-up PR to check for: `tracker.{cancel.}csum` is non-nil case.

**Pull request merged**
